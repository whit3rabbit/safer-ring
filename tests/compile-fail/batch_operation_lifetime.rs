// This test should fail to compile because batch operations must respect buffer lifetimes

use safer_ring::{Ring, Batch, Operation, PinnedBuffer};

fn main() {
    let ring = Ring::new(10).unwrap();
    let mut batch = Batch::new();
    
    {
        let mut buffer = PinnedBuffer::with_capacity(1024);
        let operation = Operation::read()
            .fd(0)
            .buffer(buffer.as_mut_slice());
        
        batch.add_operation(operation).unwrap();
    } // buffer dropped while still referenced in batch
    
    // This should not compile - buffer lifetime too short
    let _future = ring.submit_batch(batch);
}