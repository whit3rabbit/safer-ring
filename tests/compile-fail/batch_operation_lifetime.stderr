error[E0597]: `buffer` does not live long enough
  --> tests/compile-fail/batch_operation_lifetime.rs:13:21
   |
10 |         let mut buffer = PinnedBuffer::with_capacity(1024);
   |             ---------- binding `buffer` declared here
...
13 |             .buffer(buffer.as_mut_slice());
   |                     ^^^^^^ borrowed value does not live long enough
...
16 |     } // buffer dropped while still referenced in batch
   |     - `buffer` dropped here while still borrowed
...
19 |     let _future = ring.submit_batch(batch);
   |                                     ----- borrow later used here

error[E0596]: cannot borrow `ring` as mutable, as it is not declared as mutable
  --> tests/compile-fail/batch_operation_lifetime.rs:19:19
   |
19 |     let _future = ring.submit_batch(batch);
   |                   ^^^^ cannot borrow as mutable
   |
help: consider changing this to be mutable
   |
6  |     let mut ring = Ring::new(10).unwrap();
   |         +++
